pricegraph/src/orderbook/scalar.rs

```diff
@@ -169,3 +167,158 @@ fn assert_strictly_positive_and_finite(value: f64) -> f64 {
     debug_assert!(num::is_strictly_positive_and_finite(value));
     value
 }
+
+/// A signed fixed point number with 24 magnitude bits and 104 fractional bits.
+///
+/// Note the that size of the magnitude and fractional components was carefully
+/// chosen for exchange rate weights. Specifically, weights must be in a
+/// logarithmic scale so that adding them is equivalent to multiplying exchange
+/// rates. Since `log2` of the exchange rate is used, and the range of these
+/// exchange rates are:
+/// ```
+/// [MIN_AMOUNT / u128::MAX, u128::MAX / MIN_AMOUNT]
+/// ```
+///
+/// In the logarithmic scale, this range is:
+/// ```
+/// [-114.71, 114.71]
+/// ```
+///
+/// Furthermore, these weights can be added at most 2^16 times (this is the
+/// maximum number of tokens in the exchange), so the total range that must be
+/// representable by the magnitude bits is:
+/// ```
+/// [-7517784, 7517784]
```

Comment:
Not important in practice because we aren't close to a power of 2, but the `114.71` is rounded which makes a difference here.
Now that I'm writing this maybe it is easier to understand if we ignore MIN_AMOUNT.
Then the maximum log2 is `log2(2**128-1) ~ 128` (a tiny bit smaller than 128) and `128 * 2**16 == 2**23` which is exactly the 23 bits needed.

Reply:
2**23 needs 24 bits (1 with 23 0's), so the fact that the magnitude is not quite 128 is important. I also added a unit test to make sure we have enough space.

Reply:
> but the 114.71 is rounded which makes a difference here.

Yes, but only in the comment. `7517784` was computed with the original values to avoid rounding error. Note that `7517784` is the truncated value, since this is the maximum magnitude for the floating point number (it does not mean the fractional part is 0).

Reply:
Updated the comment to make more sense.